bin/tmt

```diff
@@ -9,15 +9,30 @@ try:
 
 # Show detailed output upon command execution errors
 except tmt.utils.RunError as error:
+    # Check verbosity level used during raising exception,
+    # Supported way to correctly get verbosity is
+    # tmt.util.Common.opt('verbose')
+    verbose = 0
+    tb = error.__traceback__
```

Comment:
@lukaszachy wouldn't it be simpler to provide the `tmt.utils.Common` class instance optionally in RunError when called from Common, to avoid searching for it so hard in the traceback?

Reply:
Looks to me from code, it is not raised as much as I would find this worthy

Reply:
Well, I have problem with the optionally - I'll bet lunch that at some point someone will forget to add it there, RunError will be raised and truncated.
And at the same time I don't think Common instance should be part of exception :)

This was best compromise I could found ;)

Reply:
@happz any suggestion from you here?

Reply:
To note discarded ideas because they require to different way how to raise exception
1) typeof(tmt.utils.Common) object added to the exception and use its opt('verbose') directly -> Easy to forget to set it as param and why should exception contain it if it should be discoverable from traceback frames 
2) parameter specifying verbosity level -> it has nothing to do with the exception.

If reviewers find option 1) much better than iterating through stack I will change it.

Reply:
> @happz any suggestion from you here?

No preference, can be argued in favor of many different implementations.

* optional `Common` parameter of an exception - easy, explicit.
* worry people might forget to specify? Make it mandatory then.
* "why should exception contain it if it should be discoverable from traceback frames" - it's a bit of black magic, happening implicitly behind the scenes, that's why I prefer being explicit.

Anyway, I for one prefer an explicit approach, i.e. a (possibly optional) parameter, but if you're fine with inspecting stack trace, then why not. I would only move the magic into `RunException` class, to shield its consumers from the need of digging into its traceback. Reading the code, it seems the goal is to walk the stack and discover verbosity along the way. Can we move that into some kind of property returning an iterator over all `Compose` "selfs" found in stack trace, `parents`, `owners`, `superiors`? This dissection of stack here looks a bit ad hoc-ish, digging into error internals, instead of error providing clear API to reach needed information.

Also, shouldn't there be a `max()` applied, to avoid reverting `verbose` to zero if an upper `Common` suddenly returns `verbose` lower than what has been found?

```python
def superiors(self) -> Generator['Common', None, None]:
    tb = error.__traceback__

    while tb:
        o = tb.tb_frame.f_locals.get('self', None)
        if isinstance(o, tmt.utils.Common):
            yield o

        tb = tb.tb_next

...

verbose = max(superior.opt('verbose') for superior in error.superiors, default=0)
```

Whoever then needs to find the error's origin, can do so without learning how to do it. If only the first one is important, the property can be changed to return just that instead of iterating.

Reply:
> Also, shouldn't there be a `max()` applied, to avoid reverting `verbose` to zero if an upper `Common` suddenly returns `verbose` lower than what has been found?

That is another question - should it be 'max' verbosity or verbosity of 'tmt.utils.Common' which raised the exception?
I'd went with later.

Reply:
Either way, 'explicit' is less surprising so I'll change this for `RunError` to be able to tell its verbose level directly

Reply:
> Either way, 'explicit' is less surprising so I'll change this for `RunError` to be able to tell its verbose level directly

In my books, that's another extreme position :) Why would `RunError` care about how it's presented, what should its verbosity be? Its purpose is to carry information about errors, not decide whether or not to print it out or how much of it to share with users :) That's why I draw a line in my example in serving `Common` instances known to the given error, leaving the decision on what to do with them to the caller. That leaves it up to the caller to decide whether the verbosity should be the highest one, or the first one, or the one that raised the error, or the highest one but only when logging into a file, and so on. IMHO `RunError` lacks enough context to decide what verbosity should it be presented with, because it knows nothing about its audience or channels said presentation act requires at the moment.

> > Also, shouldn't there be a `max()` applied, to avoid reverting `verbose` to zero if an upper `Common` suddenly returns `verbose` lower than what has been found?
> 
> That is another question - should it be 'max' verbosity or verbosity of 'tmt.utils.Common' which raised the exception? I'd gone with later.

True, it probably makes sense to stick to the verbosity of the origin. Still, I don't see `RunError` as the place deciding whether that's the right answer, hence I'd still propose `RunError` to provide a `origin` or similarly named property returning this particular `Common` object, leaving the verbosity management to one who knows more about the presentation.

Reply:
I like explicit, hence +1 from me.

@lukaszachy BTW yet another idea for an implementation: no extra parameter, stack inspection instead, but done in `RunError`'s `__init__()`. Peek just one frame above the current frame (`__init__()`) to find out whether there was a `Common`-like `self` object, and if so, assign it to `self.caller` (or `None`, if there was no such object). It would avoid extra parameters one might forget to specify, and implementation would be wrapped within the class itself.

(It does not go against my support for explicitness, it has nice boundaries and if needed, could be coupled with an explicit keyword parameter `caller=None` - if unset, try to detect the caller from the stack. Right now it seems `RunError` is not widely used, therefore the clear borders are slightly more important to me.)

Reply:
Nah, I'll keep it as is. Maybe next time.